src/mgr/ClusterState.cc

```diff
@@ -57,6 +57,15 @@ void ClusterState::set_mgr_map(MgrMap const &new_mgrmap)
 {
   std::lock_guard l(lock);
   mgr_map = new_mgrmap;
+  auto p = waiting_for_client_in_mgrmap.begin();
+  while (p != waiting_for_client_in_mgrmap.end()) {
+    if ((*p)->check_client_in_mgrmap(mgr_map)) {
```

Comment:
This design is a little weird relative to other code in Ceph. In particular, these checks are very specific to the register_client API.

I think it would be more appropriate to just call:

`finish_contexts(g_ceph_context, waiting_on_mgrmap);`

Then in `ActivePyModules::register_client`, create a loop which: creates the generic `C_SaferCond` and push it onto the `waiting_on_mgrmap` vector, wait for completion, then checks if the addr is in the mgrmap (success), or it should fail because the address has been removed the PyModuleRegistery::clients map, or it should loop again.

Reply:
Thanks for the explanation, Patrick. I will try this approach.